### PR TITLE
[Prometheus] Update prometheus histogram

### PR DESCRIPTION
**Pull request by constanca-m**
**Important**: All documents from the tests commited were the result of running `go test -data`, since the tests were causing this to fail.

## What

The issue is explained in https://github.com/elastic/beats/issues/36317.

To fix it, we just need to change the function `histogramMetricName` to have support for the correct suffixes:
- [x] `_count`
- [x] `_sum`
- [x] `_bucket`
- `_created` - falls in default and is later ignored.
- [x] `_gcount`
- [x] `_gsum`
- [x] everything else is ignored

The problem was occurring because we had no distinction between `*_bucket` metrics and all the others, so they were being treated as the same.

## How to test this PR locally

1. Clone this branch.
2. Build metricbeat binary and run it.

## Related issues

Closes https://github.com/elastic/beats/issues/36317.

## Results

I am running a spring application with the following metrics:
image

If we go to Discover, we can check that the documents are all there now, since the histogram is correct:
image

### CODE REVIEW

**Comment by StephanErb** [comment]
Given the discussion and linked tickets in #36317, I am wondering if we should  drop all the quantile metrics and just ingest the histogram. 

While you managed to ingest both now, you get lot of additional and expensive time series without much added value.

**Review by belimawr — APPROVED** [review]

File: metricbeat/helper/prometheus/textparse.go
Comment (ChrsMark) [inline metricbeat/helper/prometheus/textparse.go:401]: This helper is used by Prometheus and Openmetrics module at the same time. Having a link only from OpenMetric spec might be confusing here. Could you either add one for Prometheus as well and explain that it covers both or explain it specifically if it covers both?

File: metricbeat/helper/prometheus/textparse.go
Comment (ChrsMark) [inline metricbeat/helper/prometheus/textparse.go:423]: Do we have any unit-tests that cover these changes? It's hard to understand if something might break here in any corner case or whatever? 

If not would you mind adding some tests? Otherwise we can exclude these changes from the patch if not needed for the fix. Improving code should always go along with adding tests ensuring that functionality is not broken :).

File: metricbeat/helper/prometheus/textparse.go
Comment (ChrsMark) [inline metricbeat/helper/prometheus/textparse.go:571]: Is this comment no longer valid? If so why?

File: metricbeat/helper/prometheus/textparse.go
Comment (ChrsMark) [inline metricbeat/helper/prometheus/textparse.go:554]: Could you add some unit-tests for the specific use-case we are trying to fix here? This would help ensuring the introduced functionality.

**Review by ChrsMark — CHANGES_REQUESTED** [review]
Thank's @constanca-m for dealing with this.

I have some comments about the implementation details but the main one is that we will need to ensure the fix with unit tests.

Also since unit-tests might not be enough because of the importance of the this specific code-base we would need to execute some manual testing here once this is merged.
@gizas @tommyers-elastic @bturquet could we mark this one for manual testing somehow? We used to do it in the past and helped a lot. The idea is that some random member of the team (preferably not involved with the PR) would execute some manual tests to ensure that the wide range of modules affected are not broken.

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go:571]: It is, I must have removed it by accident. I will add it again.

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go:558]: It's not really a fix, I just added these lines so we could obtain the value for the quantile / le at the same time we were getting the value for labels instead of doing it later.

File: metricbeat/helper/prometheus/textparse.go
Comment (tetianakravchenko) [inline metricbeat/helper/prometheus/textparse.go:555]: it would be more readable if changed order or used `switch`

File: metricbeat/helper/prometheus/textparse.go
Comment (tetianakravchenko) [inline metricbeat/helper/prometheus/textparse.go:610]: may be worth adding a doc to the prometheus histogram definition - https://prometheus.io/docs/concepts/metric_types/#histogram

**Review by tetianakravchenko — COMMENTED** [review]
Please add a changelog record

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go]: I changed the order. Is it better now?

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go:613]: Thanks! I added it above the function used, to be side by side with OpenMetrics: https://github.com/elastic/beats/blob/561dfd74290523134f0f9e5a82d86e15d83399bc/metricbeat/helper/prometheus/textparse.go#L402.

File: metricbeat/helper/prometheus/textparse.go
Comment (ChrsMark) [reply metricbeat/helper/prometheus/textparse.go:427]: I think that for the new functionality the added test is fine but what about the existing functionality that is touched here? How we can be sure that the code which is changed (improved) is not broken?
I'm talking mostly about the changes like the one at https://github.com/elastic/beats/pull/36537/files#diff-d66f15900f5676672ddf353458a9373cf8b1ab8e237c473191e4a64ae93cbba2L407-R416.

So the point here is that if we want to improve the code we should ensure that we don't break it. So to my mind we should either:

1) ensure that we have tests for these lines that still work
2) if we don't have tests then add them along with the corresponding code changes.
3) if we cannot add tests maybe it's safer to avoid any code changes that might be prone to errors

Could you verify the above cases please?

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go:501]: Yes, when I added the test I realized that the `#TYPE ` line wasn't adding the type, and it would always be unknown.

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go:427]: We never had tests for those cases. Do you mean adding a test case for every time (histogram, histogram gauge etc)? I only changed it to `strings.TrimSuffix` because then we know what is happening instead of defining the lenght of the suffix and then subtracting like we were doing.

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go:557]: Yes, we will only end up in the `llbls.WriteString()` if both `l.Name == model.QuantileLabel` and `l.Name == labels.BucketLabel` are not true, so `if l.Name != model.QuantileLabel && l.Name != labels.BucketLabel` which was the previous condition.

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go:501]: I added a comment to explain the change.

File: metricbeat/helper/prometheus/textparse.go
Comment (ChrsMark) [reply metricbeat/helper/prometheus/textparse.go:427]: > Do you mean adding a test case for every type (histogram, histogram gauge etc)?

Ideally yes since we touch these specific lines of code. Otherwise the improvement seems to be incomplete to me and potentially prone to errors :).
Maybe that's out of the scope of this PR but the generic idea is what I mention.

File: CHANGELOG-developer.next.asciidoc
Comment (ChrsMark) [inline CHANGELOG-developer.next.asciidoc:72]: Since this is user facing message we should be more specific in what the fix is about. Sth like the following (feel free to improve more):
```suggestion
- Fix how Prometheus histograms are calculated when percentiles are provide.{pull}36537[36537]
```

File: metricbeat/helper/prometheus/textparse.go
Comment (constanca-m) [reply metricbeat/helper/prometheus/textparse.go:427]: I see. So should I try to add new tests for this PR or should we leave it to a different one?

Comment (ChrsMark) [reply metricbeat/helper/prometheus/textparse.go:427]: That would be nice since we are on this. Otherwise I doubt it will happen at any time soon.
Another option would be to exclude the specific changes/improvements that are not related to this specific fix and add them later in a follow-up along with tests. That would unblock this one and ensures the quality as well. It's up to you to decide.

**Review by ChrsMark — APPROVED** [review]
Changes lgtm. Let's make sure that we take care of the tests either here or in a follow-up based on https://github.com/elastic/beats/pull/36537#discussion_r1334209353.

Also we should consider of introducing manual testing phase for this and similar PRs that touch "core" functionality (specially when unit testing is not sufficient). @constanca-m could you please pair with @gizas and/or @tommyers-elastic on this?

**Comment by constanca-m** [comment]
Yes @ChrsMark , I am also working on adding some tests to the other types in this PR.

**Comment by constanca-m** [comment]
Actually, I will open a new PR because I already found a small error, and I don't want new changes to be added here as they are unrelated to the histogram.

**Comment by gizas** [comment]
@constanca-m please link here the additional PRs for the tests and the additional fixes

**Comment by constanca-m** [comment]
The PR with the tests is this one: https://github.com/elastic/beats/pull/36669.

**Pull request merged**
